main/src/cgeo/geocaching/connector/su/SuLogin.java

```diff
@@ -0,0 +1,186 @@
+import cgeo.geocaching.utils.Log;
+import cgeo.geocaching.utils.MatcherWrapper;
+import okhttp3.Response;
+
+public class SuLogin extends AbstractLogin {
+
+    private static final Pattern PARAMS_PATTERN_1 = Pattern.compile("oauth_token=([\\w_.-]+)");
+    private static final Pattern PARAMS_PATTERN_2 = Pattern.compile("oauth_token_secret=([\\w_.-]+)");
+    private static final String host = "www.geocaching.su";
+    private static final String FULL_HOST_URL = "http://" + host;
+    private static final String LOGON_PATH = "/api/oauth/logon.php";
+    private static final String REQUEST_TOKEN_PATH = "/api/oauth/request_token.php";
+    private static final String ACCESS_TOKEN_PATH = "/api/oauth/access_token.php";
+    private static final String AUTHORIZE_TOKEN_PATH = "/api/oauth/authorize.php";
+    private static String oAtoken;
+    private static String oAtokenSecret;
+
+    private SuLogin() {
+        // singleton
+    }
+
+    @NonNull
+    public static SuLogin getInstance() {
+        return SingletonHolder.INSTANCE;
+    }
+
+    @Override
+    @NonNull
+    protected StatusCode login(final boolean retry) {
+        return login(retry, Settings.getCredentials(SuConnector.getInstance()));
+    }
+
+    @NonNull
+    public OAuthTokens getOAuthTokens() {
+        return new OAuthTokens(oAtoken, oAtokenSecret);
+    }
+
+    @Override
+    @NonNull
+    protected StatusCode login(final boolean retry, @NonNull final Credentials credentials) {
+        if (credentials.isInvalid()) {
+            clearLoginInfo();
+            Log.w("SuLogin.login: No login information stored");
+            return StatusCode.NO_LOGIN_INFO_STORED;
+        }
+
+        final Application application = CgeoApplication.getInstance();
+        setActualStatus(application.getString(R.string.init_login_popup_working));
+
+        final Parameters params = new Parameters("username", credentials.getUserName(), "password", credentials.getPassword());
+
+        final String loginData = Network.getResponseData(Network.postRequest(FULL_HOST_URL + LOGON_PATH, params));
+
+        if (StringUtils.isBlank(loginData)) {
+            Log.e("SuLogin.login: Failed to retrieve login data");
+            setActualStatus(application.getString(R.string.init_login_popup_failed));
+            return StatusCode.CONNECTION_FAILED_SU; // no login page
+        }
+
+        if (!loginData.contains("Login succesful")) { // hardcoded in English
+            Log.i("Failed to log in Geocaching.su as " + credentials.getUserName() + " because of wrong username/password");
+            setActualStatus(application.getString(R.string.init_login_popup_failed));
+            return StatusCode.WRONG_LOGIN_DATA; // wrong login
+        }
+
+        setActualUserName(credentials.getUserName());
+
+        Log.i("Successfully logged in Geocaching.su as " + credentials.getUserName());
+        requestToken();
+        authorizeToken();
+        accessToken();
+
```

Comment:
I fail to understand how this is thought to work. First a 'traditional' login with username/password is done and than seemingly independently a pair of OAuth tokens is generated.
How are those bound to the current login?
And in general I would have expected that those tokens could be kept and used for subsequent authentications, to avoid storing the username/password in the App altogether.

Reply:
When using OAuth, the app must not see the password at all.

Reply:
This "logon" page theoretically can be called with some callback, so login data will be entered only on this page. However we (geocaching.su) don't see storing user credentials as a problem.

Tokens which we get here may have limited limetime, so it's required to request them again at some point

Reply:
> However we (geocaching.su) don't see storing user credentials as a problem.

Please reconsider that decision. It totally defeats the purpose of OAuth and is absolutely unnecessary.

> Tokens which we get here may have limited limetime, so it's required to request them again at some point

Then the API design is broken. Access tokens might have a limited lifetime, but such APIs usually provide refresh tokens that can be used to acquire a new access token without user intervention. In practice many services use only access tokens with a long or unlimited lifetime.

Reply:
done